### PR TITLE
Implement better support for MIR

### PR DESCRIPTION
**Pull request by sonmarcho**
PR Content:
===
This PR extends the support for MIR so that [Charon](https://github.com/AeneasVerif/charon) can rely on hax to extract most of the required information from the Rust compiler, instead of doing everything by itself. As such, a big portion of the code in this PR was migrated from Charon. The "bridge" between Charon and hax works: https://github.com/AeneasVerif/charon/tree/son_merge_hacspec, but there is a lot of cleaning to do on the hax side (hence this PR as a basis for discussion).

The main changes are as follows:
- The constant expressions are now merged into one `ConstantExpr` type, which is reconstructed from the various constant types manipulated by the Rust compiler. `ConstantExpr` is subsumed by `Expr`. A lot more situations are handled now. We also took the opportunity of updating the way unevaluated constants are handled to make it consistent accross the code base: unevaluated constants are evaluated if and only if they are anonymous constants, otherwise they are referenced by an id.
- The `Binder` type is closer to the Rustc `Binder` type
- I udpated the `Ty::Arrow` and added `MirFnSig` and `MirPolyFnSig`. We might want to merge those with `FnSig` and `PolyFnSig`.
- I updated the `DefId` type to include a Rust def id (needed to lookup definitions on Charon's side - see the Future Work below), and I made `DefId` generic so that we can include some type information in the `Impl` cases. Because of loop issues I had to duplicate a bit of code: it works but we need to tidy this up.

Future work:
====
Because of the work submitted in this PR, Charon now uses hax to retrieve most of the definitions coming from the Rust compiler, before translating them to (U)LLBC. By doing so, we split the work between: 1. accessing all the information we need from the Rust compiler, and 2. converting the MIR to LLBC. This allows to factor out 1. and simplify 2. as we don't do everything at once. An issue remains, which is that because we need to explore some of the external definitions Charon still interacts with the Rust compiler a lot, for example to explore the items in the crate, lookup bodies, etc. We will need to update that in the future.

**Side remark**: Charon (and Aeneas) need to access some of the external definitions because of the constraints of the symbolic execution performed by Aeneas. For instance, if some local function calls an external function, we need to retrieve the signature of this external function. In other words, we need an interface for the external crates. As of today, I compute this interface on the fly by querying exactly the public information I need about those external crates during the translation process. There are other solutions.

### CODE REVIEW

**Comment by sonmarcho** [comment]
This is joint work with @W95Psp of course

**Comment by W95Psp** [comment]
That's great, thank you @sonmarcho! :tada:
Gonna review that now

File: frontend/exporter/src/constant_utils.rs [hunk 1]
```diff
@@ -283,3 +407,104 @@ pub(crate) fn valtree_to_constant_expr<'tcx, S: BaseState<'tcx>>(
...
+    assert!(dc.variant.is_none());
+
+    // The type should be tuple
+    let hax_ty = ty.sinto(s);
+    match &hax_ty {
+        Ty::Tuple(_) => (),
+        _ => {
+            span_fatal!(s, span, "Expected the type to be tuple: {:?}", val)
+        }
+    };
```
Comment (W95Psp) [inline frontend/exporter/src/constant_utils.rs:446]: Why`hax_ty` and not `ty`? It's not used anywhere below, so we can shadow.
Also, I think using a `matches` would simply things
```suggestion
    let hax_ty = ty.sinto(s);
    if !matches(Ty::Tuple(_), &hax_ty) {
            span_fatal!(s, span, "Expected the type to be tuple: {:?}", val)
    };
```

File: frontend/exporter/src/constant_utils.rs [hunk 2]
```diff
@@ -283,3 +407,104 @@ pub(crate) fn valtree_to_constant_expr<'tcx, S: BaseState<'tcx>>(
...
+                attributes: Vec::new(),
+            }
+        }
+        ConstValue::ZeroSized { .. } => {
+            // Should be unit
+            let ty = ty.sinto(s);
+            if let Ty::Tuple(tys) = &ty && tys.is_empty() {}
+            else {
+                span_fatal!(s, span, "Expected the type to be tuple: {:?}", val)
+            }
```
Comment (W95Psp) [inline frontend/exporter/src/constant_utils.rs:499]:
```suggestion
            if !matches!(Ty::Tuple(tys), &ty) || !tys.is_empty() {
                span_fatal!(s, span, "Expected the type to be tuple: {:?}", val)
            }
```

**Review by W95Psp — COMMENTED** [review]
Some first comments
In parallel I'm fixing the rest of Hax (the engine), after the type changes.

File: frontend/exporter/src/constant_utils.rs [hunk 3]
```diff
@@ -1,11 +1,34 @@
...
+#[derive(
+    Clone, Debug, Serialize, Deserialize, JsonSchema, Hash, PartialEq, Eq, PartialOrd, Ord,
+)]
 pub enum ConstantExprKind {
-    Literal(LitKind),
+    Literal(ConstantLiteral),
     Adt {
         info: VariantInformations,
+        /// Variant index
+        vid: Option<VariantIdx>,
```
[the remark this answers is not in the document]
Comment (sonmarcho) [reply frontend/exporter/src/constant_utils.rs:31]: I was not sure where I should put it. I can move it.

File: frontend/exporter/src/constant_utils.rs [hunk 4]
```diff
@@ -176,32 +233,61 @@ pub(crate) fn translate_constant_integer_like_value<'tcx, S: BaseState<'tcx>>(
...
+/// We evaluate the anonymous constants
+pub(crate) fn must_evaluate_constant(id: &DefId) -> bool {
+    let last = id.path.last().unwrap();
+    if let DefPathItem::AnonConst = &last.data {
+        // Anonymous constant: evaluate
+        true
+    } else {
+        // Not an anonymous constant: do not evaluate
+        false
+    }
```
[the remark this answers is not in the document]
Comment (sonmarcho) [reply frontend/exporter/src/constant_utils.rs:245]: I was not aware of this notation, looks good

File: frontend/exporter/src/constant_utils.rs [hunk 5]
```diff
@@ -176,32 +233,61 @@ pub(crate) fn translate_constant_integer_like_value<'tcx, S: BaseState<'tcx>>(
...
+        if must_evaluate_constant(&id) {
+            // Anonymous constant: evaluate
+            c.eval(s.base().tcx, get_param_env(s))
+        } else {
+            // Not an anonymous constant: do not evaluate
+            c
+        }
+    } else {
+        c
+    };
```
[the remark this answers is not in the document]
Comment (sonmarcho) [reply frontend/exporter/src/constant_utils.rs:267]: And we would make `const_to_constant_expr` call itself? That would work.

File: frontend/exporter/src/constant_utils.rs (hunk 1 above)
Comment (sonmarcho) [reply frontend/exporter/src/constant_utils.rs:446]: You can change that. For the `hax_ty`: the code was ported from Charon, where I often write `rust_...`

File: frontend/exporter/src/constant_utils.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (W95Psp) [reply frontend/exporter/src/constant_utils.rs:267]: yeah, exactly :)

**Review by franziskuskiefer — APPROVED** [review]
Nice, thanks! :shipit:

**Pull request merged**